infobserve/sources/github.py

```diff
@@ -0,0 +1,112 @@
+import asyncio
+from typing import Any, Dict, List, Optional, Union
+
+import aiohttp
+
+from infobserve.common import APP_LOGGER
+from infobserve.common.index_cache import IndexCache
+from infobserve.common.queue import ProcessingQueue
+from infobserve.events import GithubEvent
+
+from .base import SourceBase
+
+BAD_CREDENTIALS = "Bad credentials"
+
+
+class GithubSource(SourceBase):
+    """The implementation of Github Source.
+
+    This Class represents the github as a source of data it fetches
+    the latest number of events specified in config and creates list of those
+    events represented as GithubEvent objects.
+
+    Attributes:
+        SOURCE_TYPE (string): The type of the source.
+        _oauth_token (string): The oauth token for the github api.
+        _username (string): The username of the user to authenticate.
+        _uri (string): Gitlab's api uri.
+        _api_version (string): Gitlab's api version.
+        _index_cache(infobserve.common.index_cache.IndexCache): IndexCache object to query the postgres cache
+        _timeout(float): The frequency the gists endpoint is queried
+        _etag(str): Returns no data if no changes detected in the api.
+    """
+
+    def __init__(self, config: Dict, name: str = None):
+        SourceBase.__init__(self, name=name)
+        self.SOURCE_TYPE: str = "github-public-events"
+        self._oauth_token: Optional[Any] = config.get('oauth')
+        self._username: Optional[Any] = config.get('username')
+        self._uri: str = "https://api.github.com/events"
+        self._api_version: str = "application/vnd.github.v3+json"
+        self._index_cache: IndexCache = IndexCache(self.SOURCE_TYPE)
+        self.timeout: Union[float] = config.get('timeout', 60)
+        self._etag: Optional[Any] = None
+
+    async def fetch_events(self) -> List[GithubEvent]:
+        """
+        Fetches the most recent gists created.
+
+        Returns:
+            event_list (list) : A list of GistEvent Objects.
+        """
+
+        headers: Dict = {
+            "User-Agent": 'Infobserver',
+            "Accept": self._api_version,
+            "Authorization": f'token {self._oauth_token}'
+        }
+
+        async with aiohttp.ClientSession(headers=headers) as session:
```

Comment:
Wait, `headers` must be set both in the session *and* in `session.get`? :open_mouth: